Better error message for quantized dispatch

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28635 Better error message for quantized dispatch**

Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

Commit:
Better error message for quantized dispatch

Fixes #28518

Commit:
Update on "Better error message for quantized dispatch"


Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

jamesr66a commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -243,7 +243,19 @@ class DispatchTable final {
-      const std::string dispatch_key_str = dispatch_key.has_value() ? toString(*dispatch_key) : "None";
+      if (dispatch_key.value() == TensorTypeId::QuantizedCPUTensorId) {
+        // If the input is quantized, but the quantization is not supported.
+        TORCH_CHECK(false, "Tried runninig '", operator_name_, "', but it",
```
s/runninig/running/

jamesr66a replied:
You might also want to specify something like "Tried running X *with a quantized tensor* but X expects a non-quantized input" just for clarity

jamesr66a commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -243,7 +243,19 @@ class DispatchTable final {
+                    " documentation for supported kernels.");
+      }
+      // If the input is not quantized, but the kernel is.
+      TORCH_CHECK(!kernels_.lookup(TensorTypeId::QuantizedCPUTensorId), "Tried",
```
For consistency, can we pull this conditional out into an `if` statement to make it clear this is not the fall-through path? I was confused when first reading this

jamesr66a reviewed (commented):
Thanks! Left some small comments

cc @smessmer

jerryzh168 commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -243,7 +243,19 @@ class DispatchTable final {
       }
 
-      const std::string dispatch_key_str = dispatch_key.has_value() ? toString(*dispatch_key) : "None";
+      if (dispatch_key.value() == TensorTypeId::QuantizedCPUTensorId) {
```
do you need to add `has_value()` check?

jamesr66a replied:
That's already checked on line 238

Commit:
Update on "Better error message for quantized dispatch"


Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

Commit:
Update on "Better error message for quantized dispatch"


Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

Commit:
Update on "Better error message for quantized dispatch"


Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

Commit:
Update on "Better error message for quantized dispatch"


Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

Commit:
Update on "Better error message for quantized dispatch"


Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

jianyuh reviewed (approved):
LGTM!

Commit:
Update on "Better error message for quantized dispatch"


Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

Commit:
Update on "Better error message for quantized dispatch"


Fixes #28518

Differential Revision: [D18132566](https://our.internmc.facebook.com/intern/diff/D18132566)

smessmer commented:
I don't think we want this. The core dispatch logic is agnostic of quantization and should stay this way. Quantization depends on dispatch, not the other way round.
